invoices/invoiceregistry.go

```diff
@@ -1019,13 +1025,62 @@ func (i *InvoiceRegistry) notifyExitHopHtlcLocked(
 	ctx *invoiceUpdateCtx, hodlChan chan<- interface{}) (
 	HtlcResolution, invoiceExpiry, error) {
 
+	invoiceRef := ctx.invoiceRef()
+	setID := (*SetID)(ctx.setID())
+
+	// We need to look up the current state of the invoice in order to
+	// allow it to be modified by the interceptor.
+	existingInvoice, err := i.idb.LookupInvoice(
+		context.Background(), invoiceRef,
+	)
+	switch {
+	case errors.Is(err, ErrInvoiceNotFound) ||
+		errors.Is(err, ErrNoInvoicesCreated):
+
+		// If this is the first HTLC for the invoice, we don't actually
+		// have the invoice in the database yet. We'll create a new and
+		// empty invoice to pass to the interceptor.
+		existingInvoice = Invoice{
+			AmtPaid: ctx.amtPaid,
+		}
+
+	case err != nil:
+		ctx.log(err.Error())
+		return nil, nil, err
+	}
+
+	// Provide the invoice to the settlement interceptor to allow
+	// the interceptor's client an opportunity to manipulate the
+	// settlement process.
+	err = i.cfg.HtlcInterceptor.Intercept(HtlcModifyRequest{
+		WireCustomRecords:  ctx.wireCustomRecords,
+		ExitHtlcCircuitKey: ctx.circuitKey,
+		ExitHtlcAmt:        ctx.amtPaid,
+		ExitHtlcExpiry:     ctx.expiry,
+		CurrentHeight:      uint32(ctx.currentHeight),
+		Invoice:            existingInvoice,
+	}, func(resp HtlcModifyResponse) {
+		log.Debugf("Received invoice HTLC interceptor response: %v",
+			resp)
+
+		if resp.AmountPaid != 0 {
+			ctx.amtPaid = resp.AmountPaid
```

Comment:
just checking: it is guaranteed that zero amt is never a valid amt here?

Reply:
on the proto side, we use an `optional` type (which compiles to a pointer) to cover this case. So maybe should do something similar here?